File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        int K = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());

        long result = 0L;
        int[] list = new int[K];

        for (int i = 0; i < K; i++) {
            list[i] = Integer.parseInt(br.readLine());
        }

        long left = 1L;
        long right = 2147483647L;

        while (left <= right) {
            long mid = (left + right) / 2;
            long cnt = 0L;
            for (int i = 0; i < K; i++) {
                cnt += list[i] / mid;
            }
            if (cnt >= N) {
                result = Math.max(result, mid);
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        bw.write(result + "");
        bw.close();
    }
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys

input = lambda: sys.stdin.readline().rstrip()

K, N = map(int, input().split())

arr = [int(input()) for _ in range(K)]


left = 1
right = 1 << 31

result = 0

while left <= right:
    mid = (left + right) // 2
    total = 0

    for a in arr:
        total += a // mid

    if total >= N:
        left = mid + 1
    else:
        right = mid - 1

print(right)
